In [1]:
import pandas as pd 
from scipy import stats

In [2]:
df=pd.read_csv("C:/Users/gupta\Documents/GitHub/Last-Mile-Delivery-Delay-Root-Cause-Analysis/data/processed/final_deliveries_clean.csv")

<>:1: SyntaxWarning: "\D" is an invalid escape sequence. Such sequences will not work in the future. Did you mean "\\D"? A raw string is also an option.
<>:1: SyntaxWarning: "\D" is an invalid escape sequence. Such sequences will not work in the future. Did you mean "\\D"? A raw string is also an option.
C:\Users\gupta\AppData\Local\Temp\ipykernel_13336\1140245213.py:1: SyntaxWarning: "\D" is an invalid escape sequence. Such sequences will not work in the future. Did you mean "\\D"? A raw string is also an option.
  df=pd.read_csv("C:/Users/gupta\Documents/GitHub/Last-Mile-Delivery-Delay-Root-Cause-Analysis/data/processed/final_deliveries_clean.csv")


In [5]:
df['Weatherconditions'].unique()


array(['conditions Sunny', 'conditions Stormy', 'conditions Sandstorms',
       'conditions Cloudy', 'conditions Fog', 'conditions Windy',
       'conditions NaN'], dtype=object)

In [4]:
df['Weatherconditions'].value_counts()

Weatherconditions
conditions Fog           7654
conditions Stormy        7586
conditions Cloudy        7536
conditions Sandstorms    7495
conditions Windy         7422
conditions Sunny         7284
conditions NaN            616
Name: count, dtype: int64

In [6]:
df['Weatherconditions'] = (df['Weatherconditions']
                           .str.replace('conditions ', '', regex=False)
                           .str.strip())
df['Weatherconditions'].unique()

array(['Sunny', 'Stormy', 'Sandstorms', 'Cloudy', 'Fog', 'Windy', 'NaN'],
      dtype=object)

In [7]:
foggy  = df[df['Weatherconditions'].isin(['Fog', 'Cloudy'])]['Time_taken(min)']
sunny  = df[df['Weatherconditions'] == 'Sunny']['Time_taken(min)']
stormy = df[df['Weatherconditions'].isin(['Stormy', 'Sandstorms'])]['Time_taken(min)']

In [11]:
print(f"Foggy rows  : {len(foggy)}")
print(f"Sunny rows  : {len(sunny)}")
print(f"Stormy rows : {len(stormy)}")

Foggy rows  : 15190
Sunny rows  : 7284
Stormy rows : 15081


In [9]:
t1, p1 =stats.ttest_ind(foggy, sunny)
print(f"\n>> Foggy/Cloudy vs ☀ Sunny")
print(f"   T-Statistic : {t1:.3f}")
print(f"   P-Value     : {p1:.5f}")
print(f"   Result      : {'✅ SIGNIFICANT!' if p1 < 0.05 else '❌ NOT significant'}")
print(f"   Foggy avg   : {foggy.mean():.1f} min")
print(f"   Sunny avg   : {sunny.mean():.1f} min")


>> Foggy/Cloudy vs ☀ Sunny
   T-Statistic : 51.781
   P-Value     : 0.00000
   Result      : ✅ SIGNIFICANT!
   Foggy avg   : 28.9 min
   Sunny avg   : 21.9 min


In [12]:
t2, p2 = stats.ttest_ind(stormy, sunny)
print(f"\n>> Stormy/Sandstorm vs ☀ Sunny")    
print(f"   T-Statistic : {t2:.3f}")
print(f"   P-Value     : {p2:.5f}")
print(f"   Result      : {'✅ SIGNIFICANT!' if p2 < 0.05 else '❌ NOT significant'}")
print(f"   Stormy avg  : {stormy.mean():.1f} min")
print(f"   Sunny avg   : {sunny.mean():.1f} min")


>> Stormy/Sandstorm vs ☀ Sunny
   T-Statistic : 33.212
   P-Value     : 0.00000
   Result      : ✅ SIGNIFICANT!
   Stormy avg  : 25.9 min
   Sunny avg   : 21.9 min


### Result:
Foggy/Cloudy weather adds +7.0 min per delivery (28.9 - 21.9)
Stormy/Sandstorm adds +4.0 min per delivery (25.9 - 21.9)
Both are statistically proven — NOT random chance.
This confirms 🌧 Environmental as a valid RCA bucket.